huaweicloud/services/dli/resource_huaweicloud_dli_datasource_connection_privilege.go

```diff
@@ -0,0 +1,248 @@
+package dli
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/jmespath/go-jmespath"
+
+	"github.com/chnsz/golangsdk"
+
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/common"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils"
+)
+
+// @API DLI PUT /v1.0/{project_id}/authorization
+// @API DLI GET /v2.0/{project_id}/datasource/enhanced-connections/{connection_id}/privileges
+func ResourceDatasourceConnectionPrivilege() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceDatasourceConnectionPrivilegeCreate,
+		UpdateContext: resourceDatasourceConnectionPrivilegeUpdate,
+		ReadContext:   resourceDatasourceConnectionPrivilegeRead,
+		DeleteContext: resourceDatasourceConnectionPrivilegeDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceDatasourceConnectionPrivilegeImportState,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"connection_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: `The ID of the connection to be granted.`,
+			},
+			"project_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: `The ID of the granted project.`,
+			},
+			"privileges": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Computed:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: `The list of permissions granted to the connection.`,
+				DiffSuppressFunc: func(_, _, _ string, d *schema.ResourceData) bool {
+					if _, ok := d.GetOk("privileges"); !ok {
+						// If the value of the permission is nil, the Computed behavior will prevent changes.
+						// If the value of the permission is not nil and the length of the array is zero, need a
+						// DiffSuppress function to prevent changes.
+						return true
+					}
+					return false
+				},
+			},
+		},
+	}
+}
+
+func buildModifyDatasourceConnectionPrivilegesBodyParams(d *schema.ResourceData, action string) map[string]interface{} {
+	// Default values of the privileges parameter is 'BIND_QUEUE'.
+	privileges := []string{"BIND_QUEUE"}
+	if d.Get("privileges").(*schema.Set).Len() > 0 {
+		privileges = utils.ExpandToStringListBySet(d.Get("privileges").(*schema.Set))
+	}
+
+	return map[string]interface{}{
+		"projectId": utils.ValueIngoreEmpty(d.Get("project_id")),
```

Comment:
project_id is required, it will never be empty

Reply:
fixed